controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1377,166 +1389,314 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+                                                             }));
+                        }));
+    }
+
+    private CompletableFuture<EpochRecord> newEpochRecord(final CompletableFuture<EpochRecord> lastRecordFuture,
+                                                          final int epoch, final int referenceEpoch,
+                                                          final Collection<StreamSegmentRecord> createdSegments,
+                                                          final Collection<Long> sealedSegments, final long time) {
+        if (epoch == referenceEpoch) {
+            return lastRecordFuture.thenApply(lastRecord -> {
+                List<StreamSegmentRecord> segments = new LinkedList<>(lastRecord.getSegments());
+                segments.removeIf(x -> sealedSegments.contains(x.segmentId()));
+                segments.addAll(createdSegments);
+                return EpochRecord.builder().epoch(epoch).referenceEpoch(referenceEpoch).segments(segments).creationTime(time).build();
+            });
+        } else {
+            return getEpochRecord(epoch);
+        }
+    }
+
+    private StreamSegmentRecord newSegmentRecord(long segmentId, long time, Double low, Double high) {
+        return newSegmentRecord(StreamSegmentNameUtils.getEpoch(segmentId), StreamSegmentNameUtils.getSegmentNumber(segmentId),
+                time, low, high);
+    }
+
+    private StreamSegmentRecord newSegmentRecord(int epoch, int segmentNumber, long time, Double low, Double high) {
+        return StreamSegmentRecord.builder().creationEpoch(epoch).segmentNumber(segmentNumber).creationTime(time)
+                                  .keyStart(low).keyEnd(high).build();
+    }
+
+    private CompletableFuture<Integer> findEpochAtTime(long timestamp) {
+        return getActiveEpoch(false)
+                .thenCompose(activeEpoch -> searchEpochAtTime(0, activeEpoch.getEpoch() / HISTORY_CHUNK_SIZE,
+                        x -> x == activeEpoch.getEpoch() / HISTORY_CHUNK_SIZE, timestamp)
+                        .thenApply(epoch -> {
+                            if (epoch == -1) {
+                                if (timestamp > activeEpoch.getCreationTime()) {
+                                    return activeEpoch.getEpoch();
+                                } else {
+                                    return 0;
+                                }
+
+                            } else {
+                                return epoch;
+                            }
+                        }));
+    }
+
+    private CompletableFuture<Integer> searchEpochAtTime(int lowest, int highest, Predicate<Integer> ignoreCached, long timestamp) {
+        final int middle = (lowest + highest) / 2;
+
+        if (lowest > highest) {
+            // either return epoch 0 or latest epoch
+            return CompletableFuture.completedFuture(-1);
+        }
+
+        return getHistoryTimeSeriesChunk(middle, ignoreCached.apply(middle))
```

Comment:
Is this doing a double binary search, like one over the chunks and another inside the chunk?

Reply:
yes. The one over chunks is performed asynchronously as we fetch the chunks from the store. 

once we have found the right chunk, its an in memory binary search though.